cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,860 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
+			if (error != HB_MC_SUCCESS) { 
+				bsg_pr_err("%s: failed to initialize tile group (%d,%d) of grid %d.\n", __func__, tg_id_x, tg_id_y, device->num_grids);
+				return HB_MC_FAIL;
+			}
+		}
+	}
+	device->num_grids ++;
+	return HB_MC_SUCCESS;
 }
 
-/*
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor, which must be obtained from hb_mc_init_host.
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Searches for a free tile group inside the device mesh and allocoates it, and sets the dimensions, origin, and id of tile group.
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_init (uint8_t *fd, eva_id_t eva_id, char *elf, tile_t *tiles, uint32_t num_tiles) {
-	
-	int error = hb_mc_fifo_init(fd); 
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_init() --> hb_mc_init_host(): failed to initialize host.\n");
-		return HB_MC_FAIL;
+int hb_mc_tile_group_allocate_tiles (device_t *device, tile_group_t *tg){
+	int error;
+	if (hb_mc_dimension_get_x(tg->dim) > hb_mc_dimension_get_x(device->mesh->dim)){
+		bsg_pr_err("%s: tile group X dimension (%d) larger than mesh X dimension (%d).\n", __func__, hb_mc_dimension_get_x(tg->dim), hb_mc_dimension_get_x(device->mesh->dim));
+		return HB_MC_INVALID;
 	}
-	
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
-	
-	for (int i = 0; i < num_tiles; i++) { /* initialize tiles */
-		hb_mc_tile_freeze_dep(*fd, tiles[i].x, tiles[i].y);
-		hb_mc_tile_set_group_origin(*fd, tiles[i].x, tiles[i].y, tiles[i].origin_x, tiles[i].origin_y);
+	if (hb_mc_dimension_get_y(tg->dim) > hb_mc_dimension_get_y(device->mesh->dim)){
+		bsg_pr_err("%s: tile group Y dimension (%d) larger than mesh Y dimension (%d).\n", __func__, hb_mc_dimension_get_y(tg->dim), hb_mc_dimension_get_y(device->mesh->dim));
+		return HB_MC_INVALID;
 	}
 
+	// Iterate over the entire mesh as tile (org_y, org_x) being the origin of the new tile group to allcoate 
+	for (hb_mc_idx_t org_y = hb_mc_coordinate_get_y(device->mesh->origin); org_y <= (hb_mc_coordinate_get_y(device->mesh->origin) + hb_mc_dimension_get_y(device->mesh->dim) - hb_mc_dimension_get_y(tg->dim)); org_y++){
+		for (hb_mc_idx_t org_x = hb_mc_coordinate_get_x(device->mesh->origin); org_x <= (hb_mc_coordinate_get_x(device->mesh->origin) + hb_mc_dimension_get_x(device->mesh->dim) - hb_mc_dimension_get_x(tg->dim)); org_x++){
+			int free = 1;
+			hb_mc_idx_t tile_id;
+			
+			// Iterate over a tg->dim.x * tg->dim.y square of tiles starting from the (org_x,org_y) checking to see if all tiles are free so they can be allocated to tile group
+			for (hb_mc_idx_t x = org_x; x < org_x + hb_mc_dimension_get_x(tg->dim); x++){
```

Comment:
This double for-loop can moved into a function like this one:
```c
bool tile_group_is_free(meh, org_x, org_y)
```

Reply:
please make it `static` -- nobody should be calling but you.

Reply:
Yes agreed, moved the double for loop into hb_mc_tiles_are_free(device, origin, dim).